### ISSUE TITLE
Add code formatting

### ISSUE THREAD

**Issue by kubawolanin** [openhab/openhab-vscode#66]
May be related to #12

TODO:

- [ ] Auto-formatting in `*.items` files (incorporate Ascii-Table?)
- [ ] Auto-formatting in `*.rules` files
- [ ] Auto-formatting in `*.sitemap` files

**Comment by kubawolanin** [comment]
In theory the Language Server does provide code formatting capabilities.
See https://github.com/openhab/openhab-vscode/issues/41#issuecomment-339247601
@sjka when my editor connects with the language server, I indeed see the  "Format Document" option in the menu. It does nothing, however, and there's nothing in the logs neither.

![image](https://user-images.githubusercontent.com/2270505/34308805-4e5eb4ce-e74f-11e7-8dbf-227f5597376b.png)

**Comment by MHerbst** [comment]
The LSP currently announces capabilities that are not implemented, see https://github.com/eclipse/smarthome/issues/6654

**Comment by SamuelBrucksch** [comment]
I would like to add this as a new feature to the newly added LSP server:
https://github.com/openhab/openhab-vscode/pull/122

**Comment by MHerbst** [comment]
> I would like to add this as a new feature to the newly added LSP serve

@SamuelBrucksch I am not sure whether we should implement the formatter in the local LSP or better in the remote LSP. For simple files like items, things (and maybe also persist and sitemaps) this would be easy locally. 
But rules can be more complex and the remote LSP that is based on XTEXT knows more about the possible syntax. Before starting an implementation I would prefer to have a closer look into the remote LSP.
cc @Confectrician

**Comment by SamuelBrucksch** [comment]
You are right, but at least for easy files like items we could start locally. Actuwlly we just need the grammar from the remote LSP then we could easily do validation and so on locally. At least for me its easier as i do not have to dive into the complexity of the java empl.

**Comment by MHerbst** [comment]
The VS Code editor provides quite a lot of useful functions but before starting to implement each the missing functions, I would prefer to have a concept where to implement which function. I would like to avoid duplicate coding. We now have the first implementation of a local LSP that can be used together with the existing remote LSP. But neither Jermome nor I have started with a review. 
I would also like to get more familiar with the remote LSP in order to discuss the best ways for the implementation of additional functions.

**Comment by SamuelBrucksch** [comment]
Sure, makes sense to avoid duplicates. Maybe we should open a seperate issue to discuss this. However even if my Java expertise is quite good i would stick with JS/TS as i'm currently more into that and i do not have too much spare time to dig into too many topics. So if its going into the java direction i have to pass.

**Comment by MHerbst** [comment]
Separate issue sounds good. Maybe we can split the work. I feel more comfortable with Java than with js/ts

**Comment by CWempe** [comment]
I guess this extension is similar to the purpose of this issue.

https://github.com/MaxBec/openHAB-Alignment-Tool

Maybe one can learn from the other.

**Comment by Confectrician** [comment]
Hey @CWempe 

@MaxBec and i had already some conversations since last year,
but we are both busy those days/months, as it seems.

But you are right, this was my intention too.
No need to reinvent the wheel.

**Comment by CWempe** [comment]
Maybe [openHAB Alignment Tool](https://marketplace.visualstudio.com/items?itemName=max-beckenbauer.oh-alignment-tool) could be added as a recommendation. (like https://github.com/openhab/openhab-vscode/issues/237#issuecomment-799762890)

**Comment by MaxBec** [comment]
Thank you @CWempe

**Comment by Confectrician** [comment]
Sure.
@MaxBec could then also recommend this extension too. 😛

**Comment by Confectrician** [comment]
Done in #242

**Comment by Confectrician** [comment]
@MaxBec small question:

I am not interested in reinventing the wheel and your extension is already pretty popular.
Are you fine with recommending each other in the extension files?

I would then like to close this issue.
For me there is no need to provide two openHAB formatter functions parallel.
I would like to use this time more efficient for solving other problems and doing different feature requests.

As already said in different messages you are highly welcome to join forces here too. 🙂

**Comment by CWempe** [comment]
From a user perspective it would make sense if both addons would merge into one addons.
I do not see a reason why a user would install the alignment addon but not the general openHAB addon.

But I leave that up to you. 😄

**Comment by MaxBec** [comment]
> @MaxBec small question:
> 
> I am not interested in reinventing the wheel and your extension is already pretty popular.
> Are you fine with recommending each other in the extension files?
> 
> I would then like to close this issue.
> For me there is no need to provide two openHAB formatter functions parallel.
> I would like to use this time more efficient for solving other problems and doing different feature requests.
> 
> As already said in different messages you are highly welcome to join forces here too. 🙂

Hey @Confectrician,

yeah of course. i have a lot of things to do at the moment. But i will do it asap!

**Comment by Confectrician** [comment]
Perfect.
I will then close this one for now. 🙂

**Comment by MaxBec** [comment]
Added the recommendation. Will be releases in 2.0.9.
https://github.com/MaxBec/openHAB-Alignment-Tool/commit/41d647af6d4b59f2801a4fee9322127bfc471a9a
